File: transports/ruby/lib/csilgen/transport/rpc.rb
# frozen_string_literal: true

module Csilgen
  module Transport
    # CSIL-RPC transport -- request/response/push envelopes -- see csil-rpc-transport.md.
    module RPC
      # A CSIL-RPC request (client -> server). `payload` is opaque CBOR(request type)
      # bytes, wrapped in tag 24 on the wire.
      RpcRequest = ::Data.define(:service, :op, :id, :payload, :auth) do
        # Data's generated initializer has no default-argument support; this override
        # supplies the optional fields' defaults (an absent id/auth is omitted on the wire).
        def initialize(service:, op:, payload: "".b, id: nil, auth: nil)
          super
        end

        def with_id(new_id) = with(id: new_id)

        def with_auth(new_auth) = with(auth: new_auth)

        def encode
          entries = [
            ["v", Conventions::VERSION],
            ["service", service],
            ["op", op],
            ["payload", Conventions.tag24(payload)]
          ]
          entries << ["id", id] unless id.nil?
          entries << ["auth", auth] unless auth.nil?
          Conventions.encode_value(Conventions.canon_map(entries))
        end

        def self.decode(data)
          value = Conventions.decode_value(data)
          Conventions.check_version(Conventions.get_uint(value, "v"))
          payload_value = Conventions.map_get(value, "payload")
          raise MalformedError, "missing 'payload'" if payload_value.nil?

          new(
            service: Conventions.get_text(value, "service"),
            op: Conventions.get_text(value, "op"),
            id: Conventions.get_uint_opt(value, "id"),
            payload: Conventions.untag24(payload_value),
            auth: Conventions.get_text_opt(value, "auth")
          )
        end
      end

      # A CSIL-RPC response (server -> client). `variant` names the declared
      # output-choice arm `payload` decodes to (the CSIL type name); `payload` is empty
      # when `status` is non-zero -- but it is still a *present* tag-24 byte string.
      RpcResponse = ::Data.define(:id, :status, :variant, :error, :payload) do
        def initialize(status:, id: nil, variant: nil, error: nil, payload: "".b)
          super
        end

        def self.ok(variant, payload)
          new(id: nil, status: Status.new(code: 0), variant: variant, error: nil, payload: payload)
        end

        def self.transport_error(status, message)
          new(id: nil, status: status, variant: nil, error: message, payload: "".b)
        end

        def with_id(new_id) = with(id: new_id)

        def encode
          entries = [
            ["v", Conventions::VERSION],
            ["status", status.code],
            ["payload", Conventions.tag24(payload)]
          ]
          entries << ["id", id] unless id.nil?
          entries << ["variant", variant] unless variant.nil?
          entries << ["error", error] unless error.nil?
          Conventions.encode_value(Conventions.canon_map(entries))
        end

        def self.decode(data)
          value = Conventions.decode_value(data)
          Conventions.check_version(Conventions.get_uint(value, "v"))
          payload_value = Conventions.map_get(value, "payload")
          payload = payload_value.nil? ? "".b : Conventions.untag24(payload_value)
          new(
            id: Conventions.get_uint_opt(value, "id"),
            status: Status.from_code(Conventions.get_int(value, "status")),
            variant: Conventions.get_text_opt(value, "variant"),
            error: Conventions.get_text_opt(value, "error"),
            payload: payload
          )
        end

        # Return self when the status is ok; otherwise raise StatusError. Callers use
        # this after decode to surface transport failures distinctly from application
        # errors (which ride as a status-0 typed variant).
        def into_transport_error
          return self if status.ok?

          raise StatusError.new(status, error)
        end
      end

      # A CSIL-RPC server push (server -> client) for `<-` operations.
      RpcPush = ::Data.define(:service, :event, :payload) do
        def initialize(service:, event:, payload: "".b)
          super
        end

        def encode
          entries = [
            ["v", Conventions::VERSION],
            ["service", service],
            ["event", event],
            ["payload", Conventions.tag24(payload)]
          ]
          Conventions.encode_value(Conventions.canon_map(entries))
        end

        def self.decode(data)
          value = Conventions.decode_value(data)
          Conventions.check_version(Conventions.get_uint(value, "v"))
          payload_value = Conventions.map_get(value, "payload")
          raise MalformedError, "missing 'payload'" if payload_value.nil?

          new(
            service: Conventions.get_text(value, "service"),
            event: Conventions.get_text(value, "event"),
            payload: Conventions.untag24(payload_value)
          )
        end
      end

      # A successful handler outcome: the chosen variant name and encoded payload.
      Reply = ::Data.define(:variant, :payload)

      # A transport-level handler outcome (no typed payload).
      TransportOutcome = ::Data.define(:status, :message)

      # A CSIL-RPC client over a frame carrier. The carrier is injected (bring your
      # own); the client owns the envelope and a per-connection monotonic correlation id.
      class Client
        attr_reader :carrier

        def initialize(carrier, multiplexed:)
          # `multiplexed` true assigns a correlation id to every request (required on
          # WS / pipelined streams); false omits it (one-in-flight carriers).
          @carrier = carrier
          @next_id = 1
          @multiplexed = multiplexed
        end

        # Invoke `service`/`op` with an encoded request payload, returning the decoded
        # response. A non-zero transport status is raised as StatusError.
        def call(service, op, payload, auth: nil)
          req = RpcRequest.new(service: service, op: op, payload: payload, auth: auth)
          if @multiplexed
            req = req.with_id(@next_id)
            @next_id += 1
          end
          @carrier.send_frame(req.encode)
          frame = @carrier.recv_frame
          raise CarrierError, "connection closed before response" if frame.nil?

          RpcResponse.decode(frame).into_transport_error
        end
      end

      # A CSIL-RPC server over a frame carrier. The host supplies a handler mapping an
      # RpcRequest to an outcome (any object responding to `call`); the generated router
      # is the natural implementation of that handler.
      class Server
        def initialize(carrier)
          @carrier = carrier
        end

        # Read one request, dispatch it through `handler`, and write the response.
        # Returns false at a clean end of stream.
        def serve_one(handler)
          frame = @carrier.recv_frame
          return false if frame.nil?

          begin
            req = RpcRequest.decode(frame)
          rescue Error => e
            # A malformed envelope cannot be correlated, so it returns transport status 1.
            resp = RpcResponse.transport_error(Status.new(code: 1), e.message)
          else
            outcome = handler.call(req)
            resp = case outcome
            in Reply(variant:, payload:)
              RpcResponse.ok(variant, payload).with_id(req.id)
            in TransportOutcome(status:, message:)
              RpcResponse.transport_error(status, message).with_id(req.id)
            end
          end
          @carrier.send_frame(resp.encode)
          true
        end
      end
    end
  end
end

File: transports/elixir/lib/csilgen/transport/rpc.ex
defmodule Csilgen.Transport.RPC do
  @moduledoc """
  CSIL-RPC transport — request/response/push envelopes — see csil-rpc-transport.md.

  Envelopes are structs; encode/decode are pure functions of binaries (no
  processes). A `null` field in a logical message is ABSENT on the wire (the key
  is omitted), and an empty response payload is still a PRESENT tag-24 byte string
  wrapping empty bytes — both are normative conventions enforced here.
  """

  alias Csilgen.Transport.{CBOR, Conventions, Status}

  defmodule Request do
    @moduledoc "A CSIL-RPC request (client → server)."
    @enforce_keys [:service, :op]
    defstruct [:service, :op, :id, :auth, payload: <<>>]

    @type t :: %__MODULE__{
            service: String.t(),
            op: String.t(),
            id: non_neg_integer() | nil,
            auth: String.t() | nil,
            payload: binary()
          }
  end

  defmodule Response do
    @moduledoc "A CSIL-RPC response (server → client)."
    @enforce_keys [:status]
    defstruct [:id, :status, :variant, :error, payload: <<>>]

    @type t :: %__MODULE__{
            id: non_neg_integer() | nil,
            status: integer(),
            variant: String.t() | nil,
            error: String.t() | nil,
            payload: binary()
          }
  end

  defmodule Push do
    @moduledoc "A CSIL-RPC server push (server → client) for `<-` operations."
    @enforce_keys [:service, :event]
    defstruct [:service, :event, payload: <<>>]

    @type t :: %__MODULE__{service: String.t(), event: String.t(), payload: binary()}
  end

  # --- Request --------------------------------------------------------------

  @doc "Builds a request with no correlation id and no per-request auth."
  @spec new_request(String.t(), String.t(), binary()) :: Request.t()
  def new_request(service, op, payload \\ <<>>) do
    %Request{service: service, op: op, payload: payload}
  end

  @spec encode_request(Request.t()) :: binary()
  def encode_request(%Request{} = r) do
    entries =
      [
        {"v", Conventions.version()},
        {"service", r.service},
        {"op", r.op},
        {"payload", Conventions.tag24(r.payload)}
      ]
      |> maybe_put("id", r.id)
      |> maybe_put("auth", r.auth)

    CBOR.encode({:map, entries})
  end

  @spec decode_request(binary()) :: {:ok, Request.t()} | {:error, term()}
  def decode_request(bytes) do
    with {:ok, map} <- Conventions.decode_envelope(bytes),
         {:ok, v} <- Conventions.get_int(map, "v"),
         :ok <- Conventions.check_version(v),
         {:ok, payload} <- Conventions.get_payload(map),
         {:ok, service} <- Conventions.get_text(map, "service"),
         {:ok, op} <- Conventions.get_text(map, "op") do
      {:ok,
       %Request{
         service: service,
         op: op,
         id: Conventions.get_uint_opt(map, "id"),
         auth: Conventions.get_text_opt(map, "auth"),
         payload: payload
       }}
    end
  end

  # --- Response -------------------------------------------------------------

  @doc "Builds a successful (status ok) typed reply."
  @spec new_response_ok(String.t(), binary()) :: Response.t()
  def new_response_ok(variant, payload) do
    %Response{status: Status.ok(), variant: variant, payload: payload}
  end

  @doc "Builds a transport-level failure (no typed payload)."
  @spec new_response_transport_error(integer(), String.t()) :: Response.t()
  def new_response_transport_error(status, message) do
    %Response{status: status, error: message, payload: <<>>}
  end

  @spec encode_response(Response.t()) :: binary()
  def encode_response(%Response{} = r) do
    entries =
      [
        {"v", Conventions.version()},
        {"status", r.status},
        {"payload", Conventions.tag24(r.payload)}
      ]
      |> maybe_put("id", r.id)
      |> maybe_put("variant", r.variant)
      |> maybe_put("error", r.error)

    CBOR.encode({:map, entries})
  end

  @spec decode_response(binary()) :: {:ok, Response.t()} | {:error, term()}
  def decode_response(bytes) do
    with {:ok, map} <- Conventions.decode_envelope(bytes),
         {:ok, v} <- Conventions.get_int(map, "v"),
         :ok <- Conventions.check_version(v),
         {:ok, payload} <- Conventions.get_payload(map),
         {:ok, status} <- Conventions.get_int(map, "status") do
      {:ok,
       %Response{
         id: Conventions.get_uint_opt(map, "id"),
         status: status,
         variant: Conventions.get_text_opt(map, "variant"),
         error: Conventions.get_text_opt(map, "error"),
         payload: payload
       }}
    end
  end

  @doc """
  Returns `{:error, {:status, name, code, message}}` for a non-ok response, or
  `:ok` when it carries a typed reply (status 0).
  """
  @spec as_transport_error(Response.t()) :: :ok | {:error, term()}
  def as_transport_error(%Response{status: status} = r) do
    if Status.ok?(status) do
      :ok
    else
      {:error, {:status, Status.name(status), status, r.error || ""}}
    end
  end

  # --- Push -----------------------------------------------------------------

  @spec new_push(String.t(), String.t(), binary()) :: Push.t()
  def new_push(service, event, payload \\ <<>>) do
    %Push{service: service, event: event, payload: payload}
  end

  @spec encode_push(Push.t()) :: binary()
  def encode_push(%Push{} = p) do
    CBOR.encode(
      {:map,
       [
         {"v", Conventions.version()},
         {"service", p.service},
         {"event", p.event},
         {"payload", Conventions.tag24(p.payload)}
       ]}
    )
  end

  @spec decode_push(binary()) :: {:ok, Push.t()} | {:error, term()}
  def decode_push(bytes) do
    with {:ok, map} <- Conventions.decode_envelope(bytes),
         {:ok, v} <- Conventions.get_int(map, "v"),
         :ok <- Conventions.check_version(v),
         {:ok, payload} <- Conventions.get_payload(map),
         {:ok, service} <- Conventions.get_text(map, "service"),
         {:ok, event} <- Conventions.get_text(map, "event") do
      {:ok, %Push{service: service, event: event, payload: payload}}
    end
  end

  # A `nil` optional field is ABSENT on the wire — the key is omitted, never encoded
  # as a present null.
  defp maybe_put(entries, _key, nil), do: entries
  defp maybe_put(entries, key, value), do: entries ++ [{key, value}]
end
